pkg/tenant/tenants.go

```diff
@@ -60,19 +75,65 @@ func (s *Server) TenantCreate(c *gin.Context) {
 	c.JSON(http.StatusCreated, out)
 }
 
-func (s *Server) TenantList(c *gin.Context) {
+// TenantDetail retrieves a summary detail of a tenant by its id and
+// returns a 200 OK response.
+//
+// Route: /tenant/:tenantID
+func (s *Server) TenantDetail(c *gin.Context) {
+	var (
+		err    error
+		tenant *api.Tenant
+	)
 
-	c.JSON(http.StatusNotImplemented, "not implemented yet")
-}
+	// Get the tenant ID from the URL and return a 404 if the
+	// tenant does not exist.
+	var tenantID ulid.ULID
+	if tenantID, err = ulid.Parse(c.Param("tenantID")); err != nil {
+		log.Debug().Err(err).Msg("could not parse tenant ulid")
+		c.JSON(http.StatusNotFound, api.ErrorResponse("could not parse tenant id"))
+		return
+	}
 
-func (s *Server) TenantDetail(c *gin.Context) {
-	c.JSON(http.StatusNotImplemented, "not implemented yet")
-}
+	// Get the specified tenant from the database and return a 500 response
+	// if it can not be retrieved.
+	if _, err = db.RetrieveTenant(c.Request.Context(), tenantID); err != nil {
+		log.Error().Err(err).Msg("could not retrieve tenant")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not retrieve tenant"))
+		return
+	}
 
-func (s *Server) TenantUpdate(c *gin.Context) {
-	c.JSON(http.StatusNotImplemented, "not implemented yet")
+	c.JSON(http.StatusOK, tenant)
 }
 
+// TenantDelete deletes a tenant from a user's request with a given
+// id and returns a 200 OK response instead of an an error response.
+//
+// Route: /tenant/:tenantID
 func (s *Server) TenantDelete(c *gin.Context) {
+	var (
+		err    error
+		tenant *api.Reply
+	)
+
+	// Get the tenant ID from the URL and return a 404 if the
+	// tenant does not exist.
+	var tenantID ulid.ULID
+	if tenantID, err = ulid.Parse(c.Param("tenantID")); err != nil {
+		log.Debug().Err(err).Msg("could not parse tenant ulid")
+		c.JSON(http.StatusNotFound, api.ErrorResponse("could not parse tenant id"))
+		return
+	}
+
+	// Delete the tenant and return a 500 response if it cannot be removed.
+	if err = db.DeleteTenant(c.Request.Context(), tenantID); err != nil {
+		log.Error().Err(err).Str("tenantID", tenantID.String()).Msg("could not delete tenant")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not delete tenant"))
+		return
+	}
+
+	c.JSON(http.StatusOK, tenant)
+}
+
+func (s *Server) TenantUpdate(c *gin.Context) {
```

Comment:
This handler will be implemented via a separate handler and this was only a position change.